cmd/gardenlet/app/gardenlet.go

```diff
@@ -0,0 +1,453 @@
+	gardencoreinformers "github.com/gardener/gardener/pkg/client/core/informers/externalversions"
+	"github.com/gardener/gardener/pkg/client/kubernetes"
+	"github.com/gardener/gardener/pkg/gardenlet/apis/config"
+	configv1alpha1 "github.com/gardener/gardener/pkg/gardenlet/apis/config/v1alpha1"
+	configvalidation "github.com/gardener/gardener/pkg/gardenlet/apis/config/validation"
+	"github.com/gardener/gardener/pkg/gardenlet/controller"
+	"github.com/gardener/gardener/pkg/gardenlet/features"
+	"github.com/gardener/gardener/pkg/logger"
+	gardenerutils "github.com/gardener/gardener/pkg/utils"
+	"github.com/gardener/gardener/pkg/version"
+
+	"github.com/sirupsen/logrus"
+	"github.com/spf13/cobra"
+	"github.com/spf13/pflag"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	"k8s.io/client-go/discovery"
+	diskcache "k8s.io/client-go/discovery/cached/disk"
+	"k8s.io/client-go/informers"
+	kubeinformers "k8s.io/client-go/informers"
+	k8s "k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/client-go/restmapper"
+	"k8s.io/client-go/tools/leaderelection"
+	"k8s.io/client-go/tools/record"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// Options has all the context and parameters needed to run a Gardenlet.
+type Options struct {
+	// ConfigFile is the location of the Gardenlet's configuration file.
+	ConfigFile string
+	config     *config.GardenletConfiguration
+	scheme     *runtime.Scheme
+	codecs     serializer.CodecFactory
+}
+
+// AddFlags adds flags for a specific Gardenlet to the specified FlagSet.
+func (o *Options) AddFlags(fs *pflag.FlagSet) {
+	fs.StringVar(&o.ConfigFile, "config", o.ConfigFile, "The path to the configuration file.")
+}
+
+// NewOptions returns a new Options object.
+func NewOptions() (*Options, error) {
+	o := &Options{
+		config: new(config.GardenletConfiguration),
+	}
+
+	o.scheme = runtime.NewScheme()
+	o.codecs = serializer.NewCodecFactory(o.scheme)
+
+	if err := config.AddToScheme(o.scheme); err != nil {
+		return nil, err
+	}
+	if err := configv1alpha1.AddToScheme(o.scheme); err != nil {
+		return nil, err
+	}
+	if err := gardencorev1alpha1.AddToScheme(scheme.Scheme); err != nil {
+		return nil, err
+	}
+
+	return o, nil
+}
+
+// loadConfigFromFile loads the contents of file and decodes it as a
+// GardenletConfiguration object.
+func (o *Options) loadConfigFromFile(file string) (*config.GardenletConfiguration, error) {
+	data, err := ioutil.ReadFile(file)
+	if err != nil {
+		return nil, err
+	}
+	return o.decodeConfig(data)
+}
+
+// decodeConfig decodes data as a GardenletConfiguration object.
+func (o *Options) decodeConfig(data []byte) (*config.GardenletConfiguration, error) {
+	configObj, gvk, err := o.codecs.UniversalDecoder().Decode(data, nil, nil)
+	if err != nil {
+		return nil, err
+	}
+	config, ok := configObj.(*config.GardenletConfiguration)
+	if !ok {
+		return nil, fmt.Errorf("got unexpected config type: %v", gvk)
+	}
```

Comment:
Why do this check, can't we simply use the Decode, then validate the error?

```suggestion
        gardenletConfig := &config.GardenletConfiguration{}
	configObj, gvk, err := o.codecs.UniversalDecoder().Decode(data, nil, gardenletConfig)
```